extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetSchemaReader.java

```diff
@@ -166,10 +187,28 @@ public static ParquetInstructions readParquetSchema(
                 return column.getPath();
             });
             final String colName;
-            final String mappedName = readInstructions.getColumnNameFromParquetColumnName(parquetColumnName);
-            if (mappedName != null) {
-                colName = mappedName;
-            } else {
+            COL_NAME: {
+                FIELD_ID: if (fieldId != null) {
+                    if (fieldIdCount.getOrDefault(fieldId.intValue(), 0L) > 1) {
+                        // This file has multiple entries for fieldId; don't match against it for field ids.
+                        break FIELD_ID;
+                    }
+                    final List<String> columnNames =
+                            readInstructions.getColumnNamesFromParquetFieldId(fieldId.intValue());
+                    if (columnNames.size() == 1) {
+                        colName = columnNames.get(0);
+                        break COL_NAME;
+                    } else if (columnNames.size() > 1) {
+                        throw new IllegalArgumentException(String.format(
+                                "Non-unique Field ID mapping provided; unable to infer TableDefinition for fieldId=%d, parquetColumnName=%s",
+                                fieldId.intValue(), parquetColumnName));
+                    }
+                }
+                final String mappedName = readInstructions.getColumnNameFromParquetColumnName(parquetColumnName);
+                if (mappedName != null) {
+                    colName = mappedName;
+                    break COL_NAME;
+                }
                 final String legalized = legalizeColumnNameFunc.apply(
```

Comment:
I think there may be a name legalization bug:
1. I think we should be using builderSupplier in the below code.
2. I think we should be recording any column we assign as a taken name, in order to ensure that we don't collide between a user-specified name and a legalized name.